text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,196 @@
+native shadow.
+
+```html
+<!-- x-parent -->
+<template>
+    <x-child>
+        <span>foo</span>
+    </x-child>
+</template>
+
+<!-- x-child -->
+<template>
+    <p>child</p>
+</template>
+```
+
+This observable difference will not be remediated as such a change would be non-trivial and
+backwards-incompatible with the current synthetic shadow implementation.
+
+### Lifecycle timing
+
+Due to the way that slotting is implemented in `@lwc/synthetic-shadow`, there is an observable
+difference in the timing of lifecycle hooks for slotted elements.
+
+As mentioned previously, in synthetic mode, slotted elements that are never assigned to a slot are
+not rendered. This means that their lifecycle hooks are never invoked.
+
+Another difference is that for synthetic mode, lifecycle hooks are invoked in the order of
+appearance after they are assigned, whereas in native mode, they are invoked in the order of
+appearance in the template.
+
+### this.shadowRoot vs this.template
+
+`this.shadowRoot` and `this.template` reference the component's shadow root in native Shadow DOM
+while `this.template` references the component's shadow root in synthetic Shadow DOM. There are no
+plans to enable `this.shadowRoot` for existing components that only support synthetic Shadow DOM as
+the current LWC-ism is to use `this.template`.
+
+### this.shadowRoot instanceof ShadowRoot
+
+Components that favor native Shadow DOM will observe that `this.shadowRoot instanceof ShadowRoot`
+evaluates to `false` in mixed mode. This is due to the fact that the synthetic shadow polyfill
+patches and replaces `ShadowRoot`. This will be remediated by customizing the `instanceof` behavior
+using `Symbol.hasInstance`.
+
+### Listening for non-composed events above the root LWC node
+
+There currently exists logic that allows listeners to handle non-composed events outside of the root
+LWC node if the event originates from a non-LWC component in the subtree. This behavior exists for
+legacy reasons and cannot be preserved for components that prefer native Shadow DOM.
+
+### Accessibility
+
+The most common accessibility issue in LWC is related to id-referencing across shadow boundaries.
+The current workaround for this is to bypass the LWC engine's id-mangling by setting attributes
+dynamically, but such a workaround would not work in native Shadow DOM.
+
+### Instrumentation
+
+Applications that rely on instrumentation libraries that don't yet support Shadow DOM are currently
+able to obtain references to descriptors like `addEventListener()` before they are patched and use
+them to override `@lwc/synthetic-shadow` polyfills where needed. Such workarounds would not work for
+events originating from components that prefer native Shadow DOM.
+
+### Styling
+
+Applications currently rely on global stylesheets to apply deeply throughout the DOM. With native
+Shadow DOM, this would not be possible. All of a component's styles would have to be generated and
+added to the component bundle.

```

Comment:
Worth mentioning here is that native shadow components will not require `lwc:dom="manual"` (as I understand it), since we won't need a MutationObserver to add the shadow token attributes for style scoping.

Reply:
This is a great point. I think we will need to keep the restriction of using `lwc:dom="manual"` even for native shadow, if components need to support both native and synthetic. We'll probably also need to list out the observable differences when manipulating DOM via DOM APIs. Thoughts?

Reply:
> I think we will need to keep the restrictions

Are you speaking about the [following restrictions](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine-core/src/framework/restrictions.ts#L102-L158)?

> We'll probably also need to list out the observable differences when manipulating DOM via DOM APIs.

The only difference between synthetic shadow + `lwc:dom=manual` and native shadow regarding styling is the timing when styles are applied. In synthetic shadow, we are using a mutation observer to track sub-tree changes to add the styling attribute. It means that in synthetic shadow DOM, styles are applied to elements a microtask later than with native shadow DOM. 

The only way I see this to break is if developers are querying for styles or layout using `Window.getComputedStyle` or `Element.prototype.getBoundingClientRect` right after they mutate the DOM but right before the mutation observer is notified. Which is quite unlikely IMO. I don't think it worth adding a section in the documentation discussion this.

Reply:
> > I think we will need to keep the restrictions

> Are you speaking about the following restrictions?

@pmdartus Yes

> It would be nice if we could remove the lwc:dom=manual requirement for native-shadow components (it's a nice carrot to encourage migration to native shadow), but I understand if we need to keep it for now because of the way the code is structured.

The problem is that without that directive, the component would work fine in native mode but throw exceptions in synthetic mode.

Reply:
Point taken. To be fair though, isn't that also the case with calling `this.shadowRoot`? (Works fine in native mode, fails in synthetic mode.)

Reply:
Thanks, added the addition of `this.shadowRoot` to synthetic mode as an open question.